Add workspace switcher

rottencandy opened the pull request:
## Fixes 
https://issues.redhat.com/browse/HAC-2747
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->


## Description
- Add workspace switcher using the ContextSwitcher component.
- Update ContextSwitcher component to support recursive menu items
- Update breadcrumb to support custom components
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->


## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 
![0](https://user-images.githubusercontent.com/20013884/216656534-9f92852b-572a-488d-8c65-ed7ecbe76946.png)
![1](https://user-images.githubusercontent.com/20013884/216656538-b043d780-5517-4191-a1b4-e6839afcf381.png)

https://user-images.githubusercontent.com/20013884/216656542-c1b39f26-629a-4338-b42f-45e724a11c9a.mp4

<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->


## How to test or reproduce?
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

<!-- - [ ] Test A -->
<!-- - [ ] Test B -->

<!-- **Test Configuration(s)**: -->


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [ ] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

rottencandy commented:
cc @openshift/team-hacdev-ux @karthikjeeyar

itsptk commented:
@rottencandy Thanks for tagging. I think we'd just want to remove the arrow ">" between the (WS) label and the workspace name, and then also replace the arrow ">" between the workspace switcher icon button and the 'proper' breadcrumb with some other separator. The design has a "/" line but we can work through whatever works.

![image](https://user-images.githubusercontent.com/54860125/217073758-06b49e48-5b44-4877-bee0-d3b7cc94fc49.png)

rottencandy commented:
@itsptk does this look good? I couldn't find that icon in PF libs so I used a slash character instead
![0](https://user-images.githubusercontent.com/20013884/217254396-4ee9ef08-bc22-4842-9959-e40b6947b1e9.png)

rottencandy commented:
/retest

itsptk commented:
> @itsptk does this look good? I couldn't find that icon in PF libs so I used a slash character instead

Thanks @rottencandy , this looks better. This might be a change to the ContextSwitcher but would it be possible to reduce the amount of space between the workspace/application name and the icon button to open the switcher? Right now there is more space between the resource and the icon button than there is the separator after it, so they sort of look disassociated. 

I think that slash looks good for now, we may take a look at another separator option and get back to you this week. Thanks again.

karthikjeeyar commented:
In the breadcrumbs `Application` contains a broken link.

<img width="438" alt="Screenshot 2023-02-08 at 12 10 18 PM" src="https://user-images.githubusercontent.com/9964343/217453912-972ceec9-d7c1-4960-948f-2ba2eb1fc7cd.png">

rottencandy commented:
Updated!
![0](https://user-images.githubusercontent.com/20013884/217492843-481534f1-ddb9-45e0-b4d1-6c4fbffc177d.png)

itsptk commented:
> Updated! ![0](https://user-images.githubusercontent.com/20013884/217492843-481534f1-ddb9-45e0-b4d1-6c4fbffc177d.png)

This looks really good, thanks @rottencandy . If possible, could we increase the space between the separator and the start of the breadcrumb (the space between "/" and "Applications" here) so that the separation between the workspace switcher and  breadcrumb is more apparent? Maybe closer matching the distance between the "..." icon button and the "/" seperator?

Also, we talked through it with the UX team more and are thinking we'd like to switch the separator from "/" to "|" to maybe make it slightly clearer. Thanks very much!

rottencandy commented:
Switched to `│` and added equal spacing to both sides, looks like this now:
![0](https://user-images.githubusercontent.com/20013884/218128608-fe4a9ccc-908c-4cee-91bd-3b4b15d3ae3d.png)

cc @karthikjeeyar

rohitkrai03 reviewed (approved):
/lgtm

rohitkrai03 commented:
/hold

rohitkrai03 commented on src/components/ApplicationDetails/ApplicationDetails.tsx:
```diff
@@ -82,14 +87,31 @@ const ApplicationDetails: React.FC<HacbsApplicationDetailsProps> = ({ applicatio
         headTitle={appDisplayName}
         breadcrumbs={[
-          { path: '/stonesoup/applications', name: 'Applications' },
+          <Badge key="badge" isRead>
```
@rottencandy Any reason to switch from breadcrumb items to breadcrumbs only?

rottencandy replied:
The `breadcrumbitems` prop would only allow adding components to the end of breadcrumbs list, and the `breadcrumbs` prop did not support custom item components.

I added support for custom components to `breadcrumbs` so we could use it to render both paths and custom items at any position in the breadcrumbs list. And we wouldn't need `breadcrumbiems` anymore.

rohitkrai03 reviewed (approved):
/lgtm

rohitkrai03 commented:
/hold cancel

We've put this feature behind MVP feature flag. It will be removed in the follow up https://github.com/openshift/hac-dev/pull/406

rottencandy commented:
/retest
